### DISCUSSION TITLE
Alt tab in Windows or Super tab in Gnome

### DISCUSSION THREAD

**Discussion by technewsguy** [sezanzeb/input-remapper#272] [Q&A]
Hello, I would like to use open the application switcher by pressing the side buttons on my Apple Mighty mouse. I tried Super_L + KEY_TAB but doesn't quite get the results I want. The task switcher comes up and rapidly switches between apps. When I do the technique on my keyboard I hold the Super key and then press tab while still holding down the Super key. Maybe it's not possible. I'm super used to doing it in macos and this would be great to do in Fedora/Gnome. Please help and thank you!

**Comment by sezanzeb** [comment]
Which user interface are you using. Gnome?

> When I do the technique on my keyboard I hold the Super key and then press enter while still holding down the Super key.

If it requires Super and enter on your keyboard, why are you mapping KEY_TAB on your mouse?

**Reply by technewsguy** [reply]
Yes, Gnome. I meant Tab. Thank you

**Reply by sezanzeb** [reply]
as of now, the only way I can think of is what was suggested by jonasBoss

**Comment by jonasBoss** [comment]
Not sure if I understand the problem correctly, but you can do this to cycle through the open windows by holding a button:
```
modify(
    Alt_L,
    hold(
        key(KEY_TAB).wait(500)
    )
)
```

**Reply by sezanzeb** [reply]
this is amazing!

**Reply by sezanzeb** [reply]
it might be interesting to be able to do something like

1. key pressed
2. super_L injected
3. tab injected
4. key pressed again
5. another tab injected
6. no key pressed within 500ms
7. super_L released

on main, a key cannot start two macros simultaneously though. I think that this would also definitely require a new macro function of some sort, can't think of a way to delay injecting super_L-up with the existing ones each time the key is pressed again.

**Reply by greecychipmunk** [reply]
Is it possible to write a macro that does this every time I press and hold a mouse button (or whatever other trigger):
1. Virtually press Alt and hold it down for as long as the mouse button is pressed.
2. Virtually press the TAB key and release it, just once.
3. Virtually release the ALT when I release the mouse button (or other trigger).

This would enable me to switch tasks using a mouse button.  If I tapped the mouse button and released it, it would switch to the previous window.  If I pressed the mouse button and held it down, it would show the task switcher, and I would be able, while pressing the mouse down, to use my arrow keys to select which application I want to switch to, and I'd switch to that as soon as I released the mouse button.  So, for as long as the mouse button is pressed, it will be like pressing ALT, but when I first press it, and before I release it, it will also virtually tap TAB and release TAB once, so as to make the task switcher panel appear.

Thank you.

**Reply by sezanzeb** [reply]
Maybe something like

```
key_down(Alt_L).
key(KEY_TAB).
hold().
key_up(Alt_L)
```

**Reply by greecychipmunk** [reply]
Thank you, but that didn't seem to work for me.  (Ubuntu 22.04).
I tried assigning it to a mouse button, to a keyboard key, and in both cases I saw nothing happen when I pressed (or held down) the assigned key.  It seems that the key_up and key_down don't do anything for me.   For instance, if I assign to a key the macro `key_down(KEY_A).key_up(KEY_A)` nothing happens.  I expected that it would be like pressing and releasing the A key, but no.  I also tried `key_down(KEY_A).hold().key_up(KEY_A)`, and still nothing.

**Reply by sezanzeb** [reply]
key_up and key_down are quite new features
